### PR TITLE
ci: incrementally update docs graph on PR merge

### PR DESCRIPTION
**Pull request by galshubeli**
## Summary

Adds a GitHub Actions workflow that triggers the FalkorDB docs knowledge-graph update on every push to `main` touching any `.md` file. The workflow itself does only three things: clone the repo, build a JSON diff payload, and POST it to GraphRAG-UI's `/api/admin/update-graph` endpoint — all SDK ingestion happens server-side.

## What the workflow does

1. Checks out this repo with full history (needed for `git diff` against `github.event.before`)
2. Runs `.github/scripts/build_diff_payload.py` to compute the `.md` diff between `BASE_SHA` (= `github.event.before`) and `HEAD_SHA` (= `github.sha`), read added + modified file contents via `git show <head>:<path>` (so historical commits work even if the working tree isn't checked out at HEAD), and write `payload.json`. Emits `skip=true` when no `.md` changes remain after filtering.
3. `curl` POSTs the payload to `${GRAPHRAG_UI_URL}/api/admin/update-graph` with a bearer token. `--fail-with-body` ensures the action exits non-zero on any non-2xx and surfaces the server's `detail` message in the log.

The endpoint does the rest: `apply_changes` + `finalize` against a UUID-suffixed copy of the live graph, smoke tests against `:Graph.questions`, atomic alias flip in the org graph, retention=1 cleanup of the previous-previous.

## Files added

- `.github/workflows/update-graph.yml` — the workflow (52 lines)
- `.github/scripts/build_diff_payload.py` — Python helper that runs inside the workflow runner; pure data transformation (git diff → JSON), no network calls, no credentials

### Repo secret (Settings → Secrets and variables → Actions → New repository secret)

- **`UPDATE_GRAPH_TOKEN`** — shared bearer token. Must match the same env var on the GraphRAG-UI deployment (Railway).

### Repo variable (Settings → Secrets and variables → Actions → Variables tab)

- **`GRAPHRAG_UI_URL`** — base URL of the GraphRAG-UI deployment, e.g. `https://staging.graphrag.falkordb.com`. No trailing slash.

That's it. **No** `FALKORDB_*`, **no** `AZURE_OPENAI_*`, **no** `GRAPHRAG_UI_CHECKOUT_PAT`. All credentials live on the GraphRAG-UI side.

## Test plan

- [x] **Server-side wet run completed on staging** (2026-05-13): manual `curl` of the same payload shape the workflow produces → endpoint accepted → `apply_changes` ran with full custom strategy parity (`SentenceTokenCapChunking(256, 2)`, `GLiNERExtractor` @ 0.75, `LLMVerifiedResolution` @ 0.95/0.80) → 14-entity schema enforced → 5 smoke-test questions passed → atomic alias flip succeeded → `:Graph.active_graph` now `docs_v6_bc0ca6b1`
- [ ] `UPDATE_GRAPH_TOKEN` set on this repo and on GraphRAG-UI's Railway env (same value)
- [ ] `GRAPHRAG_UI_URL` set as a repo variable
- [ ] Real PR merge to `main` exercises the workflow end-to-end (the remaining 10% the wet run didn't cover: trigger firing, secrets resolution, runner-environment Python execution)

## Companion PRs

- FalkorDB/GraphRAG-UI#152 — alias resolver + `/api/admin/update-graph` endpoint (merged)
- FalkorDB/GraphRAG-UI#156 — SDK 1.1.1 bump + strategy kwarg restore (merged, deployed)
- FalkorDB/GraphRAG-SDK#250 + #251 — strategy forwarding in `apply_changes`, released as v1.1.1 (both merged)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Chores**
  * Added an automated workflow to incrementally update the documentation knowledge graph on pushes to main that include markdown changes.
  * Serializes concurrent updates per branch, skips runs when no doc changes are detected, and posts a generated payload to the graph update endpoint, failing the run if the remote update does not succeed.

[Review Change Stack](https://app.coderabbit.ai/change-stack/FalkorDB/docs/pull/478)

### CODE REVIEW

File: .github/workflows/update-graph.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    branches: [main]
+    paths:
+      - "**/*.md"
+
+# Merge target is always main, so all runs share one group. Bursts of PR
+# merges queue rather than race. cancel-in-progress: false because each
+# run consumes LLM credit and we'd rather pay the wait than re-do work.
+concurrency:
+  group: update-graph-${{ github.ref }}
+  cancel-in-progress: false
```
[the remark this answers is not in the document]
Comment (galshubeli) [reply .github/workflows/update-graph.yml:18]: Real bug — agreed and fixed in `c7ae19a`. You're right that `github.ref` in a `pull_request` event resolves to `refs/pull/<N>/merge`, so each PR would have gotten its own concurrency group and they'd have run in parallel rather than queued.

Switched to `github.event.pull_request.base.ref` (target branch) so all merges to main share `update-graph-main` and serialize properly:

```yaml
concurrency:
  group: update-graph-${{ github.event.pull_request.base.ref }}
  cancel-in-progress: false
```

Server-side CAS in `/api/admin/update-graph` (FalkorDB/GraphRAG-UI#152) would have caught actual graph-state corruption from a race, but parallel runs would still have paid 2× LLM credit for what should be one ingestion. With this fix, the queue does what the intent comment promised.

File: .github/workflows/update-graph.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+          AZURE_OPENAI_API_KEY: ${{ secrets.AZURE_OPENAI_API_KEY }}
+          AZURE_OPENAI_ENDPOINT: ${{ secrets.AZURE_OPENAI_ENDPOINT }}
+          AZURE_OPENAI_DEPLOYMENT: ${{ secrets.AZURE_OPENAI_DEPLOYMENT }}
+        working-directory: graphrag-ui
+        run: |
+          python -m server.scripts.update_graph \
+            --graph-id docs_benchmark \
+            --source-root ../docs \
+            --base-sha ${{ github.event.pull_request.base.sha }} \
+            --head-sha ${{ github.sha }}
```
[the remark this answers is not in the document]
Comment (galshubeli) [reply .github/workflows/update-graph.yml]: Fixed in `c7ae19a`. Added the minimum-needed permissions block to the `update-graph` job:

```yaml
permissions:
  contents: read
```

The job only needs to read repo source (for `git diff` against the PR base) — it doesn't write back to the repo, post comments, or update statuses. The actual side effect lives on the remote GraphRAG-UI server, reached via curl + bearer token, not via `GITHUB_TOKEN`. So `contents: read` covers everything this workflow does.

Closes CodeQL alerts #14 and #15 ("Workflow does not contain permissions").

**Review by Naseem77 — CHANGES_REQUESTED** [review]
Blocking: use pull_request.merge_commit_sha instead of github.sha for HEAD_SHA

  For pull_request events, github.sha resolves to the temporary refs/pull/<N>/merge test-merge commit GitHub creates for CI, not the commit that actually lands on main. On a closed + merged == true event, the authoritative
  post-merge commit is github.event.pull_request.merge_commit_sha.

  Why this matters here:

   - The whole payload is derived from git diff $BASE_SHA $HEAD_SHA. If HEAD_SHA points at the stale test-merge SHA, the diff can drift from what actually landed on main — especially with squash/rebase merges, or if the PR was
   updated between the last test-merge and the merge button click.
   - Failure mode is silent: the workflow succeeds, the endpoint accepts the payload, and the graph is updated from the wrong file set. No alarm fires.

  Fix (one-line):

   - name: Build diff payload
     id: payload
     env:
       BASE_SHA: ${{ github.event.pull_request.base.sha }}

**Comment by galshubeli** [comment]
Confirmed and fixed in `03c58e6`. You're right — this would have silently corrupted the graph on the first real merge, especially under squash (which I'd recommended in the merge checklist). Walked through the three scenarios you raised:

- **Squash merges**: test-merge ≠ squash commit on main; tree diffs *should* line up but corner cases exist
- **Rebase merges**: definitely different commits per rebased PR commit
- **PR updated between last CI run and merge**: test-merge SHA is stale; real merge captures the newer state

All three end with the workflow happily POSTing a payload that doesn't match what's actually on main, no alarm fires. Failure mode is silent until someone notices widget answers citing wrong files.

One-line fix:

```diff
       env:
         BASE_SHA: ${{ github.event.pull_request.base.sha }}
-        HEAD_SHA: ${{ github.sha }}
+        HEAD_SHA: ${{ github.event.pull_request.merge_commit_sha }}
```

**One related subtlety worth flagging** (in the commit message too, not addressing in this PR):

`pull_request.base.sha` reflects main's tip when the PR was **last updated**, not at **merge time**. If main moves forward between those two points, our diff includes the interim landed changes too. The content-hash short-circuit in `update_graph.py` makes re-ingesting identical content a no-op, so this is wasted LLM cost rather than wrong-graph corruption.

The tighter fix would be `BASE_SHA=$(git rev-parse merge_commit_sha^1)` in a shell step, so the base is always the actual parent of the merge commit. Filing that as a v2 refinement — not blocking this PR since worst-case is cost waste on a rare timing.

**Comment by galshubeli** [comment]
**Heads-up: switched the trigger from `pull_request: closed` to `push: branches: [main]`** (commit 836fb6d).

Per product direction, "changes on main" is the source of truth rather than "PR closed." Two upsides:

1. **Catches direct pushes** to main, not just PR merges. If anyone bypasses the PR flow (rare under branch protection but possible for hotfixes or admin pushes), the graph still updates.
2. **The earlier `github.sha` vs `merge_commit_sha` quirk goes away.** On push events `github.sha` IS the actual commit on main — not a synthetic test-merge commit. `github.event.before` gives the parent. The diff payload can use both directly with no special-casing.

Net change to the YAML:

```diff
 on:
-  pull_request:
-    types: [closed]
-    branches: [main]
+  push:
+    branches:
+      - main
     paths:
       - "**/*.md"

 jobs:
   update-graph:
-    if: github.event.pull_request.merged == true   # not needed on push

 concurrency:
-  group: update-graph-${{ github.event.pull_request.base.ref }}
+  group: update-graph-${{ github.ref_name }}       # = "main"

 BASE_SHA: ${{ github.event.before }}                # parent of new HEAD
 HEAD_SHA: ${{ github.sha }}                         # new HEAD on main
```

Also added defensive handling for the all-zero `before` SHA (carried by the first push to a newly-created branch) — falls back to diffing against git's empty-tree object so the workflow doesn't crash if main is ever recreated. Edge case only.

`paths: ['**/*.md']` filter and concurrency-group serialization are preserved. CodeQL permissions block stays in place.

The previous `pull_request`-flavored fixes (concurrency target, merge_commit_sha) are obsoleted by this change but kept in history so the review trail is intact.

**Review by Naseem77 — APPROVED** [review]

File: .github/workflows/update-graph.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
+    permissions:
+      contents: read
+    env:
+      GRAPH_ID: docs_benchmark
+      GRAPHRAG_UI_URL: ${{ vars.GRAPHRAG_UI_URL }}
+    steps:
+      - name: Checkout docs
+        uses: actions/checkout@v4
```
[the remark this answers is not in the document]
Comment (galshubeli) [reply .github/workflows/update-graph.yml]: Fixed in `b745c17`. Pinned `actions/checkout` to the exact commit SHA used elsewhere in this repo (`spellcheck.yml:14`):

```yaml
uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6
```

Same convention, same SHA + trailing version comment. Coincidentally also addresses CodeRabbit's suggestion to bump past `@v4` (we're now on the v6 release that the spellcheck workflow already uses, just pinned by SHA).

File: .github/workflows/update-graph.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        with:
+          fetch-depth: 0  # full history needed for the diff below
+
+      - name: Build diff payload
+        id: payload
+        env:
+          BASE_SHA: ${{ github.event.before }}
+          HEAD_SHA: ${{ github.sha }}
+        run: python3 .github/scripts/build_diff_payload.py
```
[the remark this answers is not in the document]
Comment (galshubeli) [reply .github/workflows/update-graph.yml:40]: Updated. The PR description now describes `.github/scripts/build_diff_payload.py` as an added file and removes the "No Python is added to this repo" / "Inline Python" claims. The script ended up extracted into a tracked file (rather than inlined as a heredoc) after a refactor — easier to test locally and easier to read.

File: .github/workflows/update-graph.yml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  cancel-in-progress: false
+
+jobs:
+  update-graph:
+    if: github.event.pull_request.merged == true
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
+    steps:
+      - name: Checkout docs (this repo)
+        uses: actions/checkout@v4
```
[the remark this answers is not in the document]
Comment (galshubeli) [reply .github/workflows/update-graph.yml]: Effectively addressed by `b745c17`: pinned `actions/checkout` to the v6 commit SHA (`de0fac2e4500dabe0009e67214ff5f5447ce83dd`) that the repo's existing `spellcheck.yml` already uses, so we're past `@v4` and on a SHA-pinned reference rather than a floating tag. That's stricter than just bumping to `@v6.0.2` and matches the repo's supply-chain convention (per Copilot's separate review comment).

Skipping the `actions/setup-python` part — the workflow doesn't use that action; it relies on the runner's preinstalled `python3` for the diff-payload script. No version to bump there.
